dawn/src/interface/atlas_interface.hpp

```diff
@@ -192,6 +199,183 @@ std::vector<int> const nodeNeighboursOfNode(atlas::Mesh const& m, int const& idx
+  for(auto idx : front) {
+    for(int nbhIdx = 0; nbhIdx < table.cols(idx); nbhIdx++) {
+      newFront.push_back(table(idx, nbhIdx));
+    }
+  }
+
+  if(to == targetType) {
+    std::copy(newFront.begin(), newFront.end(), std::inserter(result, result.end()));
+  }
+
+  if(chain.size() >= 2) {
+    getNeighborsImpl(nbhTables, chain, targetType, newFront, result);
+  }
+}
+
+// entry point, kicks off the recursive function above if required
+std::vector<int> getNeighbors(atlas::Mesh const& mesh, std::vector<dawn::LocationType> chain,
+                              int idx) {
+
+  // target type is at the end of the chain (we collect all neighbors of this type "along" the
+  // chain)
+  dawn::LocationType targetType = chain.back();
+
+  // lets revert s.t. we can use the standard std::vector interface (pop_back() and back())
+  std::reverse(std::begin(chain), std::end(chain));
+
+  // convenience wrapper for all neighbor tables
+  std::unordered_map<key_t, ConnInterface, key_hash, key_equal> nbhTables;
+  nbhTables.emplace(std::make_tuple(dawn::LocationType::Cells, dawn::LocationType::Edges),
+                    ConnInterface(mesh.cells().edge_connectivity()));
+  nbhTables.emplace(std::make_tuple(dawn::LocationType::Cells, dawn::LocationType::Vertices),
+                    ConnInterface(mesh.cells().node_connectivity()));
+  nbhTables.emplace(std::make_tuple(dawn::LocationType::Edges, dawn::LocationType::Cells),
+                    ConnInterface(mesh.edges().cell_connectivity()));
+  nbhTables.emplace(std::make_tuple(dawn::LocationType::Edges, dawn::LocationType::Vertices),
+                    ConnInterface(mesh.edges().node_connectivity()));
+  nbhTables.emplace(std::make_tuple(dawn::LocationType::Vertices, dawn::LocationType::Cells),
+                    ConnInterface(mesh.nodes().cell_connectivity()));
+  nbhTables.emplace(std::make_tuple(dawn::LocationType::Vertices, dawn::LocationType::Edges),
+                    ConnInterface(mesh.nodes().edge_connectivity()));
+
+  // consume first element in chain (where we currently are, "from")
+  dawn::LocationType from = chain.back();
+  chain.pop_back();
+
+  // look at next element
+  dawn::LocationType to = chain.back();
+
+  // retrieve from->to nbh table
+  auto table = nbhTables.at({from, to});
+
+  // update the current from (the neighbors we can reach from the current index)
+  std::vector<int> front;
+  for(int nbhIdx = 0; nbhIdx < table.cols(idx); nbhIdx++) {
+    front.push_back(table(idx, nbhIdx));
+  }
+
+  // result set
+  std::set<int> result;
+
+  // if next element is of target type we collect the current front into the result
+  if(to == targetType) {
+    std::copy(front.begin(), front.end(), std::inserter(result, result.end()));
+  }
+
+  // if there are two or more elements in the chain remaining, we need to recursively keep
+  // collecting neighbors
+  if(chain.size() >= 2) {
+    getNeighborsImpl(nbhTables, chain, targetType, front, result);
+  }
```

Comment:
you can avoid the recursion and the two `if(chain.size() >= 2)` if you use `getNeighborsImpl()` as a function that adds to the result set the "selected" neighbors on a single link of the chain (e.g. middle C->V in E->**C->V-**>E->V) and you loop (here) through the chain